subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java

```diff
@@ -596,6 +597,22 @@ public void execute(TestClassResult testClassResult) {
         }
     }
 
+    private Map<String, Long> getPreviousTestClassDurations() {
+        TestResultSerializer serializer = new TestResultSerializer(getBinResultsDir());
+        if (!serializer.isHasResults()) {
+            return Collections.emptyMap();
+        }
+
+        final Map<String, Long> previousDurations = new HashMap<String, Long>();
+        serializer.read(new Action<TestClassResult>() {
```

Comment:
It would result in deserializing test results twice. First for `getPreviousFailedTestClasses`, then for `getPreviousTestClassDurations`.

I guess it would be better to deserialize just once.

Reply:
Agreed. Although I believe the deserialization takes next to no time, it looks slightly cleaner.

Reply:
It probably is proportional to the number of tests and test methods, so it probably can take a noticeable time.

Reply:
While this has since become a moot point, for due dilligence & in the spirit of Fowler's ["actually run the code and measure performance"](https://www.martinfowler.com/ieeeSoftware/yetOptimization.pdf), [I timed how long it takes](https://gist.github.com/hansi-b/302394da9b862bb0b829293aa00e7c17).